gpkit/nomials/map.py

```diff
@@ -0,0 +1,181 @@
+"Implements the NomialMap class"
+from collections import defaultdict
+import numpy as np
+from .. import DimensionalityError
+from ..small_classes import HashVector, Quantity, Strings
+from ..small_scripts import mag
+from ..varkey import VarKey
+from .substitution import parse_subs
+
+
+DIMLESS_QUANTITY = Quantity(1, "dimensionless")
+
+
+class NomialMap(HashVector):
+    "Class for efficent algebraic represention of a nomial"
+    units = None
+    expmap = None  # used for monomial-mapping postsubstitution; see .mmap()
+    csmap = None
+
+    def set_units(self, thing, thing2=None):
+        "Sets units to those of `thing*thing2`"
+        if thing is None and thing2 is None:
+            self.units = None
+        elif hasattr(thing, "units"):
+            if hasattr(thing2, "units"):
+                self.units = Quantity(1, (thing*thing2).units)
+                try:  # faster than "if self.units.dimensionless"
+                    conversion = float(self.units)
+                    self.units = None
+                    for key in self:
+                        self[key] *= conversion
+                except DimensionalityError:
+                    pass
+            elif not isinstance(thing, Quantity):
+                self.units = thing.units
+            else:
+                self.units = Quantity(1, thing.units)
+        elif hasattr(thing2, "units"):
+            self.units = Quantity(1, thing2.units)
+        else:
+            self.units = None
+
+    def to(self, units):
+        "Returns a new NomialMap of the given units"
+        sunits = self.units or DIMLESS_QUANTITY
+        nm = self * sunits.to(units).magnitude  # note that * creates a copy
+        nm.set_units(units)  # pylint: disable=no-member
+        return nm
+
+    def __add__(self, other):
+        "Adds NomialMaps together"
+        if self.units != other.units:
+            other *= float(other.units/self.units)
```

Comment:
HashVector has no awareness of units, yup. While you could create a HashVector containing units, we don't do so in Nomial